finality-aleph/src/network/address_cache.rs

```diff
@@ -0,0 +1,184 @@
+            validator_address_cache,
+            key_owner_info_provider,
+            authority_provider,
+            session_boundary_info,
+        },
+        None => ValidatorAddressCacheUpdaterImpl::Noop,
+    }
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdaterImpl<K, A> {
+    fn owner_at_session(
+        &self,
+        session: SessionId,
+        validator_index: NodeIndex,
+    ) -> Option<AccountId> {
+        match self {
+            ValidatorAddressCacheUpdaterImpl::Noop => None,
+            ValidatorAddressCacheUpdaterImpl::BackendBased {
+                key_owner_info_provider,
+                authority_provider,
+                session_boundary_info,
+                ..
+            } => {
+                let block_number = session_boundary_info
+                    .boundaries_for_session(session)
+                    .first_block();
+
+                let authority_data = authority_provider.authority_data(block_number)?;
+                let aleph_key = authority_data.authorities()[validator_index.0].clone();
+                key_owner_info_provider.aleph_key_owner(block_number, aleph_key)
+            }
+        }
+    }
+}
+
+impl<K: KeyOwnerInfoProvider, A: AuthorityProvider> ValidatorAddressCacheUpdater
+    for ValidatorAddressCacheUpdaterImpl<K, A>
+{
+    fn update(&self, validator_index: NodeIndex, info: ValidatorAddressingInfo) {
+        if let (
+            Some(validator),
+            ValidatorAddressCacheUpdaterImpl::BackendBased {
+                validator_address_cache,
+                ..
+            },
+        ) = (self.owner_at_session(info.session, validator_index), self)
+        {
+            validator_address_cache.insert(validator, info)
+        }
+    }
+}
+
+pub struct KeyOwnerInfoProviderImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: crate::aleph_primitives::AlephSessionApi<B>,
+    B: Block<Hash = BlockHash>,
+    BE: Backend<B> + 'static,
+{
+    client: Arc<C>,
+    _phantom: PhantomData<(B, BE)>,
+}
+
+impl<C, B, BE> KeyOwnerInfoProviderImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: crate::aleph_primitives::AlephSessionApi<B>,
+    B: Block<Hash = BlockHash>,
+    B::Header: Header<Number = BlockNumber>,
+    BE: Backend<B> + 'static,
+{
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<C, B, BE> KeyOwnerInfoProvider for KeyOwnerInfoProviderImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: crate::aleph_primitives::AlephSessionApi<B>,
+    B: Block<Hash = BlockHash>,
+    B::Header: Header<Number = BlockNumber>,
+    BE: Backend<B> + 'static,
+{
+    fn aleph_key_owner(&self, block_number: BlockNumber, key: AuthorityId) -> Option<AccountId> {
+        let block_hash = self.client.block_hash(block_number).ok()??;
```

Comment:
This is technically risky, as getting a block hash for an arbitrary number might give you one that is not finalized yet, and thus the data is unreliable. I'm pretty sure this is not a problem due to how this is being called, but this depends on the code in `owner_at_session`. If you add more things to this struct as I suggested in the previous comment I think this will be fine (since it is correct based on the local code), but if it remains a separate struct  please add a comment here about only running this on finalized blocks.

Reply:
After including additional structs I think we are safe as you said, because we are querying session keys from the first block of a given session, which must have been set at least in previous session.